Downloading fails, since build starts

We host a simple pypi server in our company.

This command works for most pip package, but it fails for `argon2-cffi`:

    pypi@pypihost:~> pip install  -d packages argon2-cffi
    Collecting argon2-cffi
      File was already downloaded packages/argon2_cffi-16.2.0.tar.gz
        Complete output from command python setup.py egg_info:
        Package libffi was not found in the pkg-config search path.
        Perhaps you should add the directory containing `libffi.pc'
        to the PKG_CONFIG_PATH environment variable
        No package 'libffi' found
        Package libffi was not found in the pkg-config search path.
        Perhaps you should add the directory containing `libffi.pc'
        to the PKG_CONFIG_PATH environment variable
        No package 'libffi' found
        Package libffi was not found in the pkg-config search path.
        Perhaps you should add the directory containing `libffi.pc'
        to the PKG_CONFIG_PATH environment variable
        No package 'libffi' found
        Package libffi was not found in the pkg-config search path.
        Perhaps you should add the directory containing `libffi.pc'
        to the PKG_CONFIG_PATH environment variable
        No package 'libffi' found
        Package libffi was not found in the pkg-config search path.
        Perhaps you should add the directory containing `libffi.pc'
        to the PKG_CONFIG_PATH environment variable
        No package 'libffi' found
        c/_cffi_backend.c:15:17: fatal error: ffi.h: Datei oder Verzeichnis nicht gefunden
        compilation terminated.
        Traceback (most recent call last):
          File "<string>", line 1, in <module>
          File "/home/pypi/tmp/pip-build-LJ1EjJ/argon2-cffi/setup.py", line 206, in <module>
            extras_require=EXTRAS_REQUIRE,
          File "/usr/lib/python2.7/distutils/core.py", line 112, in setup
            _setup_distribution = dist = klass(attrs)
          File "/home/pypi/lib/python2.7/site-packages/setuptools/dist.py", line 265, in __init__
            self.fetch_build_eggs(attrs['setup_requires'])
          File "/home/pypi/lib/python2.7/site-packages/setuptools/dist.py", line 310, in fetch_build_eggs
            replace_conflicting=True,
          File "/home/pypi/lib/python2.7/site-packages/pkg_resources/__init__.py", line 824, in resolve
            dist = best[req.key] = env.best_match(req, ws, installer)
          File "/home/pypi/lib/python2.7/site-packages/pkg_resources/__init__.py", line 1069, in best_match
            return self.obtain(req, installer)
          File "/home/pypi/lib/python2.7/site-packages/pkg_resources/__init__.py", line 1081, in obtain
            return installer(requirement)
          File "/home/pypi/lib/python2.7/site-packages/setuptools/dist.py", line 377, in fetch_build_egg
            return cmd.easy_install(req)
          File "/home/pypi/lib/python2.7/site-packages/setuptools/command/easy_install.py", line 629, in easy_install
            return self.install_item(spec, dist.location, tmpdir, deps)
          File "/home/pypi/lib/python2.7/site-packages/setuptools/command/easy_install.py", line 659, in install_item
            dists = self.install_eggs(spec, download, tmpdir)
          File "/home/pypi/lib/python2.7/site-packages/setuptools/command/easy_install.py", line 842, in install_eggs
            return self.build_and_install(setup_script, setup_base)
          File "/home/pypi/lib/python2.7/site-packages/setuptools/command/easy_install.py", line 1070, in build_and_install
            self.run_setup(setup_script, setup_base, args)
          File "/home/pypi/lib/python2.7/site-packages/setuptools/command/easy_install.py", line 1058, in run_setup
            raise DistutilsError("Setup script exited with %s" % (v.args[0],))
        distutils.errors.DistutilsError: Setup script exited with error: command 'gcc' failed with exit status 1
        
        ----------------------------------------
    Command "python setup.py egg_info" failed with error code 1 in /home/pypi/tmp/pip-build-LJ1EjJ/argon2-cffi/

I don't want gcc to run in this case, since I only want to **download** it.

What I am doing wrong?

Same issue on StackO: http://stackoverflow.com/questions/40038861/downloading-but-not-installing-argon2-cffi

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -11,7 +11,11 @@ The third digit is only for regressions.
 Changes:
 ^^^^^^^^
 
-*none*
+- Prevent side-effects like the installation of ``cffi`` if ``setup.py`` is called with a command that doesn't require it.
+  `#20 <https://github.com/hynek/argon2_cffi/pull/20>`_
+- Fix a bunch of warnings with new ``cffi`` versions and Python 3.6.
+  `#14 <https://github.com/hynek/argon2_cffi/pull/14>`_
+  `#16 <https://github.com/hynek/argon2_cffi/pull/16>`_
 
 
 ----

--- a/setup.py
+++ b/setup.py
@@ -1,9 +1,11 @@
 from __future__ import absolute_import, division, print_function
 
 from distutils.command.build_clib import build_clib
+from distutils.command.build import build
 from distutils.errors import DistutilsSetupError
 
 from setuptools import setup, find_packages
+from setuptools.command.install import install
 
 import sys
 import codecs
@@ -96,6 +98,132 @@ if sys.version_info[0:2] < (3, 4):
 EXTRAS_REQUIRE = {}
 
 ###############################################################################
+
+
+def keywords_with_side_effects(argv):
+    """
+    Get a dictionary with setup keywords that (can) have side effects.
+
+    :param argv: A list of strings with command line arguments.
+
+    :returns: A dictionary with keyword arguments for the ``setup()`` function.
+        This setup.py script uses the setuptools 'setup_requires' feature
+        because this is required by the cffi package to compile extension
+        modules. The purpose of ``keywords_with_side_effects()`` is to avoid
+        triggering the cffi build process as a result of setup.py invocations
+        that don't need the cffi module to be built (setup.py serves the dual
+        purpose of exposing package metadata).
+
+    Stolen from pyca/cryptography.
+    """
+    no_setup_requires_arguments = (
+        '-h', '--help',
+        '-n', '--dry-run',
+        '-q', '--quiet',
+        '-v', '--verbose',
+        '-V', '--version',
+        '--author',
+        '--author-email',
+        '--classifiers',
+        '--contact',
+        '--contact-email',
+        '--description',
+        '--egg-base',
+        '--fullname',
+        '--help-commands',
+        '--keywords',
+        '--licence',
+        '--license',
+        '--long-description',
+        '--maintainer',
+        '--maintainer-email',
+        '--name',
+        '--no-user-cfg',
+        '--obsoletes',
+        '--platforms',
+        '--provides',
+        '--requires',
+        '--url',
+        'clean',
+        'egg_info',
+        'register',
+        'sdist',
+        'upload',
+    )
+
+    def is_short_option(argument):
+        """Check whether a command line argument is a short option."""
+        return len(argument) >= 2 and argument[0] == '-' and argument[1] != '-'
+
+    def expand_short_options(argument):
+        """Expand combined short options into canonical short options."""
+        return ('-' + char for char in argument[1:])
+
+    def argument_without_setup_requirements(argv, i):
+        """Check whether a command line argument needs setup requirements."""
+        if argv[i] in no_setup_requires_arguments:
+            # Simple case: An argument which is either an option or a command
+            # which doesn't need setup requirements.
+            return True
+        elif (is_short_option(argv[i]) and
+              all(option in no_setup_requires_arguments
+                  for option in expand_short_options(argv[i]))):
+            # Not so simple case: Combined short options none of which need
+            # setup requirements.
+            return True
+        elif argv[i - 1:i] == ['--egg-base']:
+            # Tricky case: --egg-info takes an argument which should not make
+            # us use setup_requires (defeating the purpose of this code).
+            return True
+        else:
+            return False
+
+    if all(argument_without_setup_requirements(argv, i)
+           for i in range(1, len(argv))):
+        return {
+            "cmdclass": {
+                "build": DummyBuild,
+                "install": DummyInstall,
+            }
+        }
+    else:
+        return {
+            "setup_requires": SETUP_REQUIRES,
+            "cffi_modules": CFFI_MODULES,
+            "libraries": LIBRARIES,
+            "cmdclass": {
+                "build_clib": BuildCLibWithCompilerFlags,
+            },
+        }
+
+
+setup_requires_error = (
+    "Requested setup command that needs 'setup_requires' while command line "
+    "arguments implied a side effect free command or option."
+)
+
+
+class DummyBuild(build):
+    """
+    This class makes it very obvious when ``keywords_with_side_effects()`` has
+    incorrectly interpreted the command line arguments to ``setup.py build`` as
+    one of the 'side effect free' commands or options.
+    """
+
+    def run(self):
+        raise RuntimeError(setup_requires_error)
+
+
+class DummyInstall(install):
+    """
+    This class makes it very obvious when ``keywords_with_side_effects()`` has
+    incorrectly interpreted the command line arguments to ``setup.py install``
+    as one of the 'side effect free' commands or options.
+    """
+
+    def run(self):
+        raise RuntimeError(setup_requires_error)
+
 
 HERE = os.path.abspath(os.path.dirname(__file__))
 
@@ -195,13 +323,12 @@ if __name__ == "__main__":
         keywords=KEYWORDS,
         packages=PACKAGES,
         package_dir={"": "src"},
-        cffi_modules=CFFI_MODULES,
-        ext_package="argon2",
-        libraries=LIBRARIES,
-        cmdclass={"build_clib": BuildCLibWithCompilerFlags},
-        zip_safe=False,
         classifiers=CLASSIFIERS,
-        setup_requires=SETUP_REQUIRES,
         install_requires=INSTALL_REQUIRES,
         extras_require=EXTRAS_REQUIRE,
+
+        # CFFI
+        zip_safe=False,
+        ext_package="argon2",
+        **keywords_with_side_effects(sys.argv)
     )

